// @brillout/json-serializer/dist/cjs/stringify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isJsonSerializerError = exports.stringify = void 0;
const types_1 = require("./types");
const isReactElement_1 = require("./utils/isReactElement");
const isCallable_1 = require("./utils/isCallable");
const isObject_1 = require("./utils/isObject");
const replacerWithPath_1 = require("./utils/replacerWithPath");
function stringify(value, { forbidReactElements, space, valueName, sortObjectKeys } = {}) {
    // The only error `JSON.stringify()` can throw is `TypeError "cyclic object value"`.
    // - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#exceptions
    // - This means we have total of 3 possible errors while serializing:
    //    - Cyclic references
    //    - Functions
    //    - React elements
    const serializer = (val) => JSON.stringify(val, (0, replacerWithPath_1.replacerWithPath)(replacer, !valueName), space);
    return serializer(value);
    function replacer(key, value, path) {
        if (forbidReactElements && (0, isReactElement_1.isReactElement)(value)) {
            throw genErr(genErrMsg('React element', path, valueName));
        }
        if ((0, isCallable_1.isCallable)(value)) {
            const functionName = value.name;
            throw genErr(genErrMsg('function', path, valueName, path.length === 0 ? functionName : undefined));
        }
        const valueOriginal = this[key];
        for (const { is, serialize } of types_1.types.slice().reverse()) {
            if (is(valueOriginal)) {
                //@ts-ignore
                return serialize(valueOriginal, serializer);
            }
        }
        if (sortObjectKeys && (0, isObject_1.isObject)(value)) {
            const copy = {};
            Object.keys(value)
                .sort()
                .forEach((key) => {
                copy[key] = value[key];
            });
            value = copy;
        }
        return value;
    }
}
exports.stringify = stringify;
const stamp = '_isJsonSerializerError';
function genErr(errMsg) {
    const err = new Error(`[@brillout/json-serializer](https://github.com/brillout/json-serializer) ${errMsg}.`);
    Object.assign(err, {
        messageCore: errMsg,
        [stamp]: true
    });
    return err;
}
function isJsonSerializerError(thing) {
    return (0, isObject_1.isObject)(thing) && thing[stamp] === true;
}
exports.isJsonSerializerError = isJsonSerializerError;
function genErrMsg(valueType, path, rootValueName, problematicValueName) {
    let subject;
    if (!path) {
        subject = rootValueName || problematicValueName || 'value';
    }
    else {
        if (problematicValueName) {
            subject = problematicValueName + ' at ';
        }
        else {
            subject = '';
        }
        subject = subject + (rootValueName || '') + path;
    }
    return `cannot serialize ${subject} because it's a ${valueType}`;
}
