// vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/replaceImportStatements.js
export { replaceImportStatements };
export { parseImportData };
export { isImportData };
// Playground: https://github.com/brillout/acorn-playground
import { parse } from 'acorn';
import { assert, assertUsage, assertWarning, styleFileRE } from '../../../utils.js';
import pc from '@brillout/picocolors';
function replaceImportStatements(code, filePathToShowToUser) {
    const spliceOperations = [];
    const fileImports = [];
    const imports = getImports(code);
    if (imports.length === 0)
        return { noImportStatement: true };
    imports.forEach((node) => {
        if (node.type !== 'ImportDeclaration')
            return;
        const importPath = node.source.value;
        assert(typeof importPath === 'string');
        const { start, end } = node;
        const importStatementCode = code.slice(start, end);
        // No variable imported
        if (node.specifiers.length === 0) {
            const isWarning = !styleFileRE.test(importPath);
            let quote = indent(importStatementCode);
            if (isWarning) {
                quote = pc.cyan(quote);
            }
            else {
                quote = pc.bold(pc.red(quote));
            }
            const errMsg = [
                `As explained in https://vike.dev/header-file the following import in ${filePathToShowToUser} has no effect:`,
                quote
            ].join('\n');
            if (!isWarning) {
                assertUsage(false, errMsg);
            }
            assertWarning(false, errMsg, { onlyOnce: true });
        }
        let replacement = '';
        node.specifiers.forEach((specifier) => {
            assert(specifier.type === 'ImportSpecifier' ||
                specifier.type === 'ImportDefaultSpecifier' ||
                specifier.type === 'ImportNamespaceSpecifier');
            const importLocalName = specifier.local.name;
            const exportName = (() => {
                if (specifier.type === 'ImportDefaultSpecifier')
                    return 'default';
                if (specifier.type === 'ImportNamespaceSpecifier')
                    return '*';
                {
                    const imported = specifier.imported;
                    if (imported)
                        return imported.name;
                }
                return importLocalName;
            })();
            const importString = serializeImportData({ importPath, exportName, importStringWasGenerated: true });
            replacement += `const ${importLocalName} = '${importString}';`;
            fileImports.push({
                importStatementCode,
                importString,
                importLocalName
            });
        });
        spliceOperations.push({
            start,
            end,
            replacement
        });
    });
    const codeMod = spliceMany(code, spliceOperations);
    return { code: codeMod, fileImports, noImportStatement: false };
}
function getImports(code) {
    const { body } = parse(code, {
        ecmaVersion: 'latest',
        sourceType: 'module'
        // https://github.com/acornjs/acorn/issues/1136
    });
    const imports = [];
    body.forEach((node) => {
        if (node.type === 'ImportDeclaration')
            imports.push(node);
    });
    return imports;
}
const import_ = 'import';
const SEP = ':';
const zeroWidthSpace = '\u200b';
function serializeImportData({ importPath, exportName, importStringWasGenerated }) {
    const tag = importStringWasGenerated ? zeroWidthSpace : '';
    // `import:${importPath}:${importPath}`
    return `${tag}${import_}${SEP}${importPath}${SEP}${exportName}`;
}
function isImportData(str) {
    return str.startsWith(import_ + SEP) || str.startsWith(zeroWidthSpace + import_ + SEP);
}
function parseImportData(importString) {
    if (!isImportData(importString)) {
        return null;
    }
    let importStringWasGenerated = false;
    if (importString.startsWith(zeroWidthSpace)) {
        importStringWasGenerated = true;
        assert(zeroWidthSpace.length === 1);
        importString = importString.slice(1);
    }
    const parts = importString.split(SEP).slice(1);
    if (!importStringWasGenerated && parts.length === 1) {
        const exportName = 'default';
        const importPath = parts[0];
        return { importPath, exportName, importStringWasGenerated, importString };
    }
    assert(parts.length >= 2);
    const exportName = parts[parts.length - 1];
    const importPath = parts.slice(0, -1).join(SEP);
    return { importPath, exportName, importStringWasGenerated, importString };
}
function spliceMany(str, operations) {
    let strMod = '';
    let endPrev;
    operations.forEach(({ start, end, replacement }) => {
        if (endPrev !== undefined) {
            assert(endPrev < start);
        }
        else {
            endPrev = 0;
        }
        const replaced = str.slice(start, end);
        strMod +=
            str.slice(endPrev, start) +
                replacement +
                // Preserve number of lines for source map
                Array(replaced.split('\n').length - replacement.split('\n').length)
                    .fill('\n')
                    .join('');
        endPrev = end;
    });
    strMod += str.slice(endPrev, str.length - 1);
    return strMod;
}
function indent(str) {
    return str
        .split('\n')
        .map((s) => `  ${s}`)
        .join('\n');
}
