// vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/getVikeConfig/configDefinitionsBuiltIn.js
export { configDefinitionsBuiltIn };
export { configDefinitionsBuiltInGlobal };
import { assert } from '../../../../utils.js';
const configDefinitionsBuiltIn = {
    onRenderHtml: {
        env: { server: true }
    },
    onRenderClient: {
        env: { client: true }
    },
    onHydrationEnd: {
        env: { client: true }
    },
    onPageTransitionStart: {
        env: { client: true }
    },
    onPageTransitionEnd: {
        env: { client: true }
    },
    onBeforeRender: {
        env: { server: true }
    },
    onBeforePrerenderStart: {
        env: { server: true }
    },
    Page: {
        env: { server: true, client: true }
    },
    passToClient: {
        env: { server: true, config: true },
        cumulative: true
    },
    route: {
        env: { server: true, client: 'if-client-routing', eager: true }
    },
    guard: {
        env: { server: true, client: 'if-client-routing' }
    },
    data: {
        env: { server: true }
    },
    iKnowThePerformanceRisksOfAsyncRouteFunctions: {
        env: { server: true, client: 'if-client-routing', eager: true }
    },
    filesystemRoutingRoot: {
        env: { config: true }
    },
    client: {
        // The value of the client config is merely the file path to the client entry file, which is only needed on the sever-side
        env: { server: true },
        _valueIsFilePath: true
    },
    clientRouting: {
        // We could make it { client: false } but we don't yet because of some legacy V0.4 design code
        env: { server: true, client: true, config: true, eager: true }
    },
    prerender: {
        env: { config: true }
    },
    hydrationCanBeAborted: {
        env: { client: true }
    },
    prefetchStaticAssets: {
        env: { client: true }
    },
    extends: {
        env: { config: true }
    },
    meta: {
        env: { config: true }
    },
    isClientSideRenderable: {
        env: { server: true, client: true, eager: true },
        _computed: (configValueSources) => isConfigSet(configValueSources, 'onRenderClient') &&
            isConfigSet(configValueSources, 'Page') &&
            !!getConfigEnv(configValueSources, 'Page')?.client
    },
    onBeforeRenderEnv: {
        env: { client: true },
        _computed: (configValueSources) => !isConfigSet(configValueSources, 'onBeforeRender') ? null : getConfigEnv(configValueSources, 'onBeforeRender')
    },
    dataEnv: {
        env: { client: true },
        _computed: (configValueSources) => !isConfigSet(configValueSources, 'data') ? null : getConfigEnv(configValueSources, 'data')
    },
    hooksTimeout: {
        env: { server: true, client: true }
    }
};
const configDefinitionsBuiltInGlobal = {
    onPrerenderStart: {
        env: { server: true }
    },
    onBeforeRoute: {
        env: { server: true, client: 'if-client-routing', eager: true }
    },
    prerender: {
        env: { config: true }
    },
    extensions: { env: { config: true } },
    disableAutoFullBuild: { env: { config: true } },
    includeAssetsImportedByServer: { env: { config: true } },
    baseAssets: { env: { config: true } },
    baseServer: { env: { config: true } },
    redirects: { env: { server: true } },
    trailingSlash: { env: { server: true } },
    disableUrlNormalization: { env: { server: true } }
};
function getConfigEnv(configValueSources, configName) {
    const configValueSource = getConfigValueSource(configValueSources, configName);
    if (!configValueSource)
        return null;
    return configValueSource.configEnv;
}
function isConfigSet(configValueSources, configName) {
    const configValueSource = getConfigValueSource(configValueSources, configName);
    // Enable users to suppress global config values by overriding the config's value to null
    if (configValueSource?.value === null)
        return false;
    return !!configValueSource;
}
function getConfigValueSource(configValueSources, configName) {
    const sources = configValueSources[configName];
    if (!sources)
        return null;
    const configValueSource = sources[0];
    assert(configValueSource);
    return configValueSource;
}
