// vike/dist/esm/shared/page-configs/helpers.js
export { getPageConfig };
export { getConfigValueFilePathToShowToUser };
export { getHookFilePathToShowToUser };
export { getConfigDefinedAtString, getDefinedAtString } from './helpers/getConfigDefinedAtString.js';
export { getConfigValue } from './helpers/getConfigValue.js';
import { assert } from '../utils.js';
function getPageConfig(pageId, pageConfigs) {
    const pageConfig = pageConfigs.find((p) => p.pageId === pageId);
    assert(pageConfigs.length > 0);
    assert(pageConfig);
    return pageConfig;
}
function getConfigValueFilePathToShowToUser({ definedAt }) {
    // A unique file path only exists if the config value isn't cumulative nor computed:
    //  - cumulative config values have multiple file paths
    //  - computed values don't have any file path
    if ('isComputed' in definedAt || 'files' in definedAt)
        return null;
    const { filePathToShowToUser } = definedAt;
    assert(filePathToShowToUser);
    return filePathToShowToUser;
}
function getHookFilePathToShowToUser({ definedAt }) {
    const filePathToShowToUser = getConfigValueFilePathToShowToUser({ definedAt });
    assert(filePathToShowToUser);
    return filePathToShowToUser;
}
