// vike/dist/esm/client/client-routing-runtime/history.js
export { initHistoryState, getHistoryState, pushHistory, saveScrollPosition, monkeyPatchHistoryPushState };
import { assert, assertUsage, getGlobalObject, hasProp, isObject } from './utils.js';
const globalObject = getGlobalObject('history.ts', {});
// Fill missing state information:
//  - `history.state` can uninitialized (i.e. `null`):
//    - The very first render
//    - The user's code runs `location.hash = '#section'`
//    - The user clicks on an anchor link `<a href="#section">Section</a>` (Vike's `onLinkClick()` handler skips hash links).
//  - State information may be incomplete if `history.state` is set by an old Vike version. (E.g. `state.timestamp` was introduced for `pageContext.isBackwardNavigation` in `0.4.19`.)
function initHistoryState() {
    let state = window.history.state;
    if (!state) {
        state = {};
    }
    let hasModifications = false;
    if (!('timestamp' in state)) {
        hasModifications = true;
        state.timestamp = getTimestamp();
    }
    if (!('scrollPosition' in state)) {
        hasModifications = true;
        state.scrollPosition = getScrollPosition();
    }
    if (!('triggedBy' in state)) {
        state.triggedBy = 'browser';
    }
    assertState(state);
    if (hasModifications) {
        replaceHistoryState(state);
    }
}
function getHistoryState() {
    const state = window.history.state || {};
    assertState(state);
    return state;
}
function getScrollPosition() {
    const scrollPosition = { x: window.scrollX, y: window.scrollY };
    return scrollPosition;
}
function getTimestamp() {
    return new Date().getTime();
}
function saveScrollPosition() {
    const scrollPosition = getScrollPosition();
    const state = getHistoryState();
    replaceHistoryState({ ...state, scrollPosition });
}
function pushHistory(url, overwriteLastHistoryEntry) {
    if (!overwriteLastHistoryEntry) {
        const timestamp = getTimestamp();
        pushHistoryState({ timestamp, scrollPosition: null, triggedBy: 'vike' }, url);
    }
    else {
        replaceHistoryState(getHistoryState(), url);
    }
}
function assertState(state) {
    assert(isObject(state));
    if ('timestamp' in state) {
        const { timestamp } = state;
        assert(typeof timestamp === 'number');
    }
    if ('scrollPosition' in state) {
        const { scrollPosition } = state;
        if (scrollPosition !== null) {
            assert(hasProp(scrollPosition, 'x', 'number') && hasProp(scrollPosition, 'y', 'number'));
        }
    }
}
function replaceHistoryState(state, url) {
    window.history.replaceState(state, '', url ?? /* Passing `undefined` chokes older Edge versions */ null);
}
function pushHistoryState(state, url) {
    pushStateOriginal(state, '', url);
}
function monkeyPatchHistoryPushState() {
    globalObject.pushStateOriginal = globalObject.pushStateOriginal ?? window.history.pushState;
    window.history.pushState = (stateFromUser = {}, ...rest) => {
        assertUsage(null === stateFromUser || undefined === stateFromUser || isObject(stateFromUser), 'history.pushState(state) argument state must be an object');
        const state = {
            scrollPosition: getScrollPosition(),
            timestamp: getTimestamp(),
            ...stateFromUser,
            // Don't allow user to overwrite triggedBy as it would break Vike's handling of the 'popstate' event
            triggedBy: 'user'
        };
        return pushStateOriginal(state, ...rest);
    };
}
function pushStateOriginal(...args) {
    globalObject.pushStateOriginal.apply(history, args);
}
